Track app time as a new metric

carlosantoniodasilva opened the pull request:
This can be useful to help determine internally if queue time increase
has direct relation to app time increase, or actual capacity issues, so
we'll track it as a new metric.

carlosantoniodasilva committed:
Track app time as a new metric

This can be useful to help determine internally if queue time increase
has direct relation to app time increase, or actual capacity issues, so
we'll track it as a new metric.

adamlogic commented on judoscale-ruby/lib/judoscale/request_middleware.rb:
```diff
@@ -33,9 +33,16 @@ def call(env)
         logger.debug "Request queue_time=#{queue_time}ms network_time=#{network_time}ms request_id=#{request_metrics.request_id} size=#{request_metrics.size}"
-      end
 
-      @app.call(env)
```
This implementation has us only tracking app time if `request_metrics.track?` is true, which certainly makes sense in terms of the method name, but in practice it means app time tracking is dependent on the request start header, and I'm not sure that's right.

`RequestMetrics#track` is a misleading name—it's specifically telling us whether we're able to reliably track request queue time, not whether we can track _any_ request metrics.

I can't think of a reason to skip _app time_, even if queue time isn't available.

carlosantoniodasilva replied:
In my mind they were meant to go together, i.e. we'd want app time if we have queue time, but it's probably okay to track app time even if we don't have queue time available.

adamlogic replied:
Here's one scenario where I think it'd be helpful: An ECS customer has installed the adapter but has not set up the NGINX sidecar. There would be no queue time, but we'd still have app time. The customer wouldn't see app time, but if they reached out to support, this extra information would show us that everything is set up correct except NGINX.

Side note... I'm just now realizing that if utilization-based scaling works out, the NGINX sidecar on ECS might be optional! 🥳

carlosantoniodasilva replied:
Right, I thought the reports could be enough in some cases like this, but I could see app time metrics could be useful as well, I'm making the change.

> I'm just now realizing that if utilization-based scaling works out, the NGINX sidecar on ECS might be optional

Nice! 🎉

carlosantoniodasilva committed:
Always track app time, regardless of queue time availability

We want to always track & report app time, regardless of having queue
time available (or skipping it due to large payloads), so that it can
help with debugging in some cases. They don't need to be linked to each
other for reporting, i.e. no need to skip app time because we are not
tracking queue time.

adamlogic reviewed (approved):
Love it!
